crypto/rsa/rsa_backend.c

```diff
@@ -97,3 +105,152 @@ int rsa_fromdata(RSA *rsa, const OSSL_PARAM params[])
+        goto err;
+    ret = 1;
+ err:
+    sk_BIGNUM_const_free(factors);
+    sk_BIGNUM_const_free(exps);
+    sk_BIGNUM_const_free(coeffs);
+    return ret;
+}
+
+int rsa_pss_params_30_todata(const RSA_PSS_PARAMS_30 *pss, const char *propq,
+                             OSSL_PARAM_BLD *bld, OSSL_PARAM params[])
+{
+    const char *key_md = OSSL_PKEY_PARAM_RSA_DIGEST;
+    const char *mdname = rsa_oaeppss_nid2name(rsa_pss_params_30_hashalg(pss));
+    const char *key_mgf = OSSL_PKEY_PARAM_RSA_MASKGENFUNC;
+    const char *mgfname = rsa_mgf_nid2name(NID_mgf1);
+    const char *key_mgf1_md = OSSL_PKEY_PARAM_RSA_MGF1_DIGEST;
+    const char *mgf1mdname =
+        rsa_oaeppss_nid2name(rsa_pss_params_30_maskgenalg(pss));
+    const char *key_saltlen = OSSL_PKEY_PARAM_RSA_SALTLEN;
+    int min_saltlen = rsa_pss_params_30_saltlen(pss);
+
+    if (mdname == NULL || mgfname == NULL || mgf1mdname == NULL
+        || !ossl_param_build_set_utf8_string(bld, params, key_md, mdname)
+        || !ossl_param_build_set_utf8_string(bld, params, key_mgf, mgfname)
+        || !ossl_param_build_set_utf8_string(bld, params, key_mgf1_md,
+                                             mgf1mdname)
+        || !ossl_param_build_set_int(bld, params, key_saltlen, min_saltlen))
+        return 0;
+    return 1;
+}
+
+int rsa_pss_params_30_fromdata(RSA_PSS_PARAMS_30 **rsa_pss_params,
+                               const OSSL_PARAM params[], OPENSSL_CTX *libctx)
+{
+    const OSSL_PARAM *param_md, *param_mgf, *param_mgf1md,  *param_saltlen;
+    char mdname[OSSL_MAX_ALGORITHM_ID_SIZE] = "", *pmdname = mdname;
+    char mgf1mdname[OSSL_MAX_ALGORITHM_ID_SIZE] = "", *pmgf1mdname = mgf1mdname;
+    EVP_MD *md, *mgf1md;
+    int saltlen;
+    RSA_PSS_PARAMS_30 *pss_params = NULL, *alloc_pss_params = NULL;
+    int ret = 0;
+
+    if (rsa_pss_params == NULL)
+        return 0;
+
+    param_md =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_DIGEST);
+    param_mgf =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_MASKGENFUNC);
+    param_mgf1md =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_MGF1_DIGEST);
+    param_saltlen =
+        OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_RSA_SALTLEN);
+
+    if (param_md == NULL && param_mgf1md == NULL && param_saltlen == NULL)
+        return 1;
+
+    if (param_mgf != NULL) {
+        const char *mgfname = NULL;
+
+        if (param_mgf->data_type == OSSL_PARAM_UTF8_STRING)
+            mgfname = param_mgf->data;
+        else if (!OSSL_PARAM_get_utf8_ptr(param_mgf, &mgfname))
+            return 0;
+        if (strcasecmp(param_mgf->data, rsa_mgf_nid2name(NID_mgf1)) != 0)
+            return 0;
+    }
+
+    if (*rsa_pss_params == NULL)
+        pss_params = alloc_pss_params = rsa_pss_params_30_new();
+    else
+        pss_params = *rsa_pss_params;
+    if (pss_params == NULL)
+        return 0;
+
+    if ((param_md != NULL && !OSSL_PARAM_get_utf8_string(param_md, &pmdname,
```

Comment:
I think I prefer checking the type and pointing to the string data directly without having to copy to a temporary local buffer for these.. I wonder if there should be an API variant to do this common case..

Reply:
Done